play-restli/src/main/java/com/linkedin/playrestli/BaseRestliServerComponent.java

```diff
@@ -20,40 +19,34 @@
  */
 public abstract class BaseRestliServerComponent<T extends Request> {
   protected final HttpDispatcher _restliDispatcher;
+  private final ClientCookieEncoder _cookieEncoder;
 
-  protected BaseRestliServerComponent(HttpDispatcher restliDispatcher) {
+  protected BaseRestliServerComponent(HttpDispatcher restliDispatcher, CookiesConfiguration cookiesConfiguration) {
     _restliDispatcher = restliDispatcher;
-  }
-
-  /**
-   * Transform the request headers to a Map by ignoring multiple values.
-   */
-  protected Map<String, String> toSimpleMap(Map<String, List<String>> map) {
-    Map<String, String> result = new HashMap<>();
-    for (Map.Entry<String, List<String>> entry : map.entrySet()) {
-      String commaSeparatedValues = StringUtils.join(entry.getValue(), ",");
-      result.put(entry.getKey(), commaSeparatedValues);
-    }
-    return result;
+    _cookieEncoder = cookiesConfiguration.clientEncoder();
   }
 
   protected <B extends BaseRequestBuilder<B>> B createRequestBuilder(
       Http.Request request, Function<URI, B> createBuilder) throws Exception {
+    B builder = createBuilder.apply(new URI(request.uri()));
+    builder.setMethod(request.method());
 
+    for (Map.Entry<String, List<String>> header : request.getHeaders().toMap().entrySet()) {
+      String key = header.getKey();
 
-    B builder = createBuilder.apply(new URI(request.uri()));
-    Map<String, List<String>> headers = request.getHeaders().toMap();
+      if (key.equalsIgnoreCase(Http.HeaderNames.COOKIE)) {
+        continue; // Cookie header and request.cookies may be out of sync; request.cookies is the source of truth.
+      }
+
+      for (String value : header.getValue()) {
+        builder.addHeaderValue(key, value);
+      }
+    }
+
+    for (Http.Cookie cookie : request.cookies()) {
+      builder.addCookie(_cookieEncoder.encode(cookie.name(), cookie.value()));
+    }
```

Comment:
To save some lines here, you could write something like `builder.setCookies(Streams.stream(request.cookies()).map(cookie -> encoder.encode(new DefaultCookie(cookie.name(), cookie.value()))).collect(
        Collectors.toList())`.

Unfortunately list of cookies to list of encoded strings is not supported. :/